cmd/agent/app/run.go

```diff
@@ -81,6 +84,19 @@ func run(cmd *cobra.Command, args []string) error {
 		}
 	}()
 
+	// By default systemd redirect the stdout to journald. When journald is stopped or crashes we receive a SIGPIPE signal.
+	// Go ignore SIGPIPE signals unless it is when stdout or stdout is closed, in this case the agent is stopped. We don't want that
+	// so we intercept the SIGPIPE signals and just discard them.
+	if ignoreSigpipe {
```

Comment:
any reason not to use [`signal.Ignore`](https://golang.org/pkg/os/signal/#Ignore) instead of this custom logic?

Reply:
`signal.Ignore` only undoes the effect of `signal.Notify`. The doc is not very clear about it but I've tried this first

Reply:
I was amazed by this assumption and had to try it out. My intention was not to test the validity of your statement, but to actually open an issue on [golang/go](https://github.com/golang/go) to correct the documentation. However, when trying it out, it seems that what you say is not true. Feel free to try it out yourself with this simple program:

```go
package main

import (
	"os/signal"
	"syscall"
	"time"
)

func main() {
	signal.Ignore(syscall.SIGINT)
	time.Sleep(5 * time.Second)
}
```
You can experiment with and without the `Ignore` call.

Reply:
SIGPIPE is a special case for go (it even has its own section in the documentation here: https://golang.org/pkg/os/signal/). The documentation also suggests to use `Notify` for what we want to do. I'm building a package of the agent using Ignore just to be sure. We can find it on gitlab on the branch remicalixte/sigpipe once the pipeline is finished. I will test in with journald since I don't know of any other setup where you can get a `SIGPIPE` from stdout without killing your process.

Reply:
I can confirm `Ignore` is not enough
```
vagrant@ubuntu-xenial:/etc/datadog-agent$ sudo systemctl status datadog-agent
● datadog-agent.service - "Datadog Agent"
   Loaded: loaded (/lib/systemd/system/datadog-agent.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2018-10-12 14:44:18 UTC; 3s ago
 Main PID: 2051 (agent)
    Tasks: 10
   Memory: 20.5M
      CPU: 510ms
   CGroup: /system.slice/datadog-agent.service
           └─2051 /opt/datadog-agent/bin/agent/agent run -p /opt/datadog-agent/run/agent.pid

Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:76 in Enter) | Scheduling check memory with an interval of 15s
Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:63 in Schedule) | Scheduling check network
Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:76 in Enter) | Scheduling check network with an interval of 15s
Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:63 in Schedule) | Scheduling check ntp
Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:76 in Enter) | Scheduling check ntp with an interval of 15s
Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:63 in Schedule) | Scheduling check uptime
Oct 12 14:44:19 ubuntu-xenial agent[2051]: 2018-10-12 14:44:19 UTC | INFO | (scheduler.go:76 in Enter) | Scheduling check uptime with an interval of 15s
Oct 12 14:44:20 ubuntu-xenial agent[2051]: 2018-10-12 14:44:20 UTC | INFO | (runner.go:258 in work) | Running check cpu
Oct 12 14:44:20 ubuntu-xenial agent[2051]: 2018-10-12 14:44:20 UTC | INFO | (runner.go:324 in work) | Done running check cpu
Oct 12 14:44:21 ubuntu-xenial agent[2051]: 2018-10-12 14:44:21 UTC | INFO | (serializer.go:242 in SendMetadata) | Sent host metadata payload, size: 3359 bytes.
vagrant@ubuntu-xenial:/etc/datadog-agent$ sudo systemctl stop systemd-journald
Warning: Stopping systemd-journald.service, but it can still be activated by:
  systemd-journald-dev-log.socket
  systemd-journald.socket
  systemd-journald-audit.socket
vagrant@ubuntu-xenial:/etc/datadog-agent$ sudo systemctl status datadog-agent
● datadog-agent.service - "Datadog Agent"
   Loaded: loaded (/lib/systemd/system/datadog-agent.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Fri 2018-10-12 14:44:39 UTC; 2s ago
  Process: 2051 ExecStart=/opt/datadog-agent/bin/agent/agent run -p /opt/datadog-agent/run/agent.pid (code=killed, signal=PIPE)
 Main PID: 2051 (code=killed, signal=PIPE)

Oct 12 14:44:31 ubuntu-xenial agent[2051]: 2018-10-12 14:44:31 UTC | INFO | (runner.go:324 in work) | Done running check uptime
Oct 12 14:44:32 ubuntu-xenial agent[2051]: 2018-10-12 14:44:32 UTC | INFO | (runner.go:258 in work) | Running check network
Oct 12 14:44:32 ubuntu-xenial agent[2051]: 2018-10-12 14:44:32 UTC | WARN | (datadog_agent.go:149 in LogMessage) | (base.py:223) | DEPRECATION NOTICE: `device_name` is deprecated, please use a `device:` tag in the `tags` list instead
Oct 12 14:44:32 ubuntu-xenial agent[2051]: 2018-10-12 14:44:32 UTC | INFO | (runner.go:324 in work) | Done running check network
Oct 12 14:44:33 ubuntu-xenial agent[2051]: 2018-10-12 14:44:33 UTC | INFO | (runner.go:258 in work) | Running check load
Oct 12 14:44:33 ubuntu-xenial agent[2051]: 2018-10-12 14:44:33 UTC | INFO | (runner.go:324 in work) | Done running check load
Oct 12 14:44:34 ubuntu-xenial agent[2051]: 2018-10-12 14:44:34 UTC | INFO | (runner.go:258 in work) | Running check file_handle
Oct 12 14:44:34 ubuntu-xenial agent[2051]: 2018-10-12 14:44:34 UTC | INFO | (runner.go:324 in work) | Done running check file_handle
Oct 12 14:44:35 ubuntu-xenial agent[2051]: 2018-10-12 14:44:35 UTC | INFO | (runner.go:258 in work) | Running check cpu
Oct 12 14:44:35 ubuntu-xenial agent[2051]: 2018-10-12 14:44:35 UTC | INFO | (runner.go:324 in work) | Done running check cpu
vagrant@ubuntu-xenial:/etc/datadog-agent$
```
- agent is running
- we kill journald
- agent exited with a SIGPIPE

Reply:
Thanks for clarifying this.